src/Symfony/Component/OptionsResolver/OptionsResolver.php

```diff
@@ -784,15 +786,43 @@ public function resolve(array $options = array())
      * @throws OptionDefinitionException If there is a cyclic dependency between
      *                                   lazy options and/or normalizers
      */
-    public function offsetGet($option)
+    public function offsetGet($option/*, bool $triggerDeprecation = true*/)
     {
         if (!$this->locked) {
             throw new AccessException('Array access is only supported within closures of lazy options and normalizers.');
         }
 
+        $triggerDeprecation = 1 === \func_num_args() || \func_get_arg(1);
+
         // Shortcut for resolved options
         if (array_key_exists($option, $this->resolved)) {
-            return $this->resolved[$option];
+            $value = $this->resolved[$option];
+
+            if (isset($this->deprecated[$option]) && $triggerDeprecation) {
+                $deprecationMessage = $this->deprecated[$option];
+
+                if ($deprecationMessage instanceof \Closure) {
+                    // If the closure is already being called, we have a cyclic dependency
+                    if (isset($this->calling[$option])) {
+                        throw new OptionDefinitionException(sprintf('The options "%s" have a cyclic dependency.', implode('", "', array_keys($this->calling))));
+                    }
+
+                    $this->calling[$option] = true;
+                    try {
+                        if (!\is_string($deprecationMessage = $deprecationMessage($this, $value))) {
```

Comment:
The deprecation closure could be called twice, the 1st time (below) with a non-normalized value, the 2nd time (here) with a normalized value, after normalization the value could change, so this does not look good :(

See alternative https://github.com/symfony/symfony/pull/28878

Reply:
i expected each time you call `Options $options['some']` it should trigger, thus each time you consume it (whether it's resolved already or not).

Reply:
unless it should not trigger of course :) (i.e. in core, hence the new `$triggerDeprecation` arg)

Reply:
My concern is more about the value that is being passed to the closure func, it must be the raw value after validating it and at this point the value is already normalized.

Reply:
i see, do we need that? Is there any reason to vary a deprecation message based on value? Isnt it deprecated by definition anyway

if so, we can move the trigger to the top (before the early return) in terms of DRY code.

Reply:
maybe yes :) it allows to deprecate a specific value instead of the option as a whole. Neat i guess :sweat_smile: 

What about storing the raw resolved values as well then? To be passed when resolved.